cv32e40s/tests/programs/custom/illegal_access_loop_tests/illegal_access_loop_tests.c

```diff
@@ -0,0 +1,89 @@
+/*
+**
+** Copyright 2022 OpenHW Group
+**
+** Licensed under the Solderpad Hardware Licence, Version 2.0 (the "License");
+** you may not use this file except in compliance with the License.
+** You may obtain a copy of the License at
+**
+** https://solderpad.org/licenses/
+**
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+**
+*******************************************************************************
+**
+** Contains generated tests on U-mode access and privilege instructions.
+**
+*******************************************************************************
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include "corev_uvmt.h"
+#include <stdint.h>
+
+// extern and global variable declaration
+extern volatile void  setup_pmp();
+extern volatile unsigned int illegal_full(), csr_loop(), mco_loop(), Check_mcounteren();
+
+
+// Declaration of assert 
+static __inline__ void assert_or_die(uint32_t actual, uint32_t expect, char *msg) {
+  if (actual != expect) {
+    printf(msg);
+    printf("expected = 0x%lx (%ld), got = 0x%lx (%ld)\n", expect, (int32_t)expect, actual, (int32_t)actual);
+    exit(EXIT_FAILURE);
+  }
+}
+
+void change_m_mode(void) {
+  //s4 is x20
+  __asm__ volatile("addi s4, x0, 0x3"); // add 3 to the s4 register, which will be read by the trap handler late
+  __asm__ volatile("ecall");
+  __asm__ volatile("li s4, 0"); // Reset the value after entering machine mode
+}
+
+void illegal_custom_loop(void) {
+    /* 
+    Tests U-mode access to various custom functions which are not yet implemented. Should all trap.
+    */
+    change_m_mode();
+    volatile unsigned int epp = 0;
+    //setup_pmp();
+    epp = illegal_full();
+    assert_or_die(epp, 131072, "error: not all illegal custom instructions triggered the trap handler\n");
```

Comment:
What is `131072`?
For so called "magic numbers" it is nice for readability to define a named constant.
(Although, admittedly we don't always.)
Alternatively an explanatory comment is also a nice way to show what the number comes from.

Reply:
131072 is the number of illegal custom instructions. The Python file will also print the number of lines (the relevant number for the assert function) after writing.